src/subtype.c

```diff
@@ -1177,22 +1177,221 @@ JL_DLLEXPORT int jl_subtype_env_size(jl_value_t *t)
+            return 1;
+        return 0;
+    }
+    if (x == (jl_value_t*)jl_bottom_type) {
+        *subtype = 1;
+        return 1;
+    }
+    if (y == (jl_value_t*)jl_bottom_type) {
+        *subtype = 0;
+        return 1;
+    }
+    if (!jl_is_type(x) || !jl_is_type(y)) {
+        *subtype = jl_egal(x, y);
+        return 1;
+    }
+    if (jl_is_uniontype(x)) {
+        // TODO: consider handling more LHS unions, being wary of covariance
+        if (jl_obvious_subtype(((jl_uniontype_t*)x)->a, y, subtype) && *subtype) {
+            if (jl_obvious_subtype(((jl_uniontype_t*)x)->b, y, subtype) && *subtype)
+                return 1;
+        }
+        //if (jl_obvious_subtype(((jl_uniontype_t*)x)->a, y, subtype)) {
+        //    if (!*subtype)
+        //        return 1;
+        //    if (jl_obvious_subtype(((jl_uniontype_t*)x)->b, y, subtype))
+        //        return 1;
+        //}
+        //else if (jl_obvious_subtype(((jl_uniontype_t*)x)->b, y, subtype)) {
+        //    if (!*subtype)
+        //        return 1;
+        //}
+        return 0;
+    }
+    if (jl_is_uniontype(y)) {
+        if (jl_obvious_subtype(x, ((jl_uniontype_t*)y)->a, subtype)) {
+            if (*subtype)
+                return 1;
+            if (jl_obvious_subtype(x, ((jl_uniontype_t*)y)->b, subtype))
+                return 1;
+        }
+        else if (jl_obvious_subtype(x, ((jl_uniontype_t*)y)->b, subtype)) {
+            if (*subtype)
+                return 1;
+        }
+        return 0;
+    }
+    if (x == (jl_value_t*)jl_any_type) {
+        *subtype = 0;
+        return 1;
+    }
+    if (jl_is_datatype(y)) {
+        int istuple = (((jl_datatype_t*)y)->name == jl_tuple_typename);
+        int iscov = istuple || (((jl_datatype_t*)y)->name == jl_vararg_typename);
+        // TODO: this would be a nice fast-path to have, unfortuanately,
+        //       datatype allocation fails to correctly cons them
+        //       and the subtyping tests include tests for this case
+        //if (!iscov && ((jl_datatype_t*)y)->isconcretetype && !jl_is_type_type(x)) {
+        //    *subtype = 0;
+        //    return 1;
+        //}
+        if (jl_is_datatype(x)) {
+            int uncertain = 0;
+            if (((jl_datatype_t*)x)->name != ((jl_datatype_t*)y)->name) {
+                if (jl_is_type_type(x) || jl_is_type_type(y))
+                    return 0;
+                jl_datatype_t *temp = (jl_datatype_t*)x;
+                while (temp->name != ((jl_datatype_t*)y)->name) {
+                    temp = temp->super;
+                    if (temp == NULL) // invalid state during type declaration
+                        return 0;
+                    if (temp == jl_any_type) {
+                        *subtype = 0;
+                        return 1;
+                    }
+                }
+                if (jl_obvious_subtype((jl_value_t*)temp, y, subtype) && *subtype)
+                    return 1;
+                return 0;
```

Comment:
Couldn't this be simplified to
```c
return (jl_obvious_subtype((jl_value_t*)temp, y, subtype) && *subtype)
```
?

Reply:
yes, but it would break the symmetry with the rest of the function